pygmt/clib/session.py

```diff
@@ -1791,19 +1802,19 @@ def virtualfile_in(  # noqa: PLR0912
             "vectors": self.virtualfile_from_vectors,
         }[kind]
 
-        # Ensure the data is an iterable (Python list or tuple).
+        # "_data" is the data that will be passed to the _virtualfile_from function.
+        # "_data" defaults to "data" but should be adjusted for some cases.
+        _data = data
         match kind:
-            case "arg" | "file" | "geojson" | "grid" | "image" | "stringio":
```

Comment:
Maybe just put `_data = data` in the `case _` below? I'd actually prefer if we make it more explicit like:

```
case "arg" | "file" | "geojson" | "grid" | "stringio":
    _data = data
```

So that if there's a new `kind` in the future, we know to add it explicitly.

Reply:
> I'd actually prefer if we make it more explicit like:
> 
> ```
> case "arg" | "file" | "geojson" | "grid" | "stringio":
>     _data = data
> ```
> 
> So that if there's a new `kind` in the future, we know to add it explicitly.

We can't do it like this, because we're currently writing case statements like `case "image" if data.dtype != "uint8":`, which only catches the case when `kind="image"` and `data.dtype != "uint8"`.

We can refactor the codes to:
```
case "arg" | "file" | "geojson" | "grid" | "stringio":
    _data = data
case "image":
    _data = data
    if data.dtype != "uint8":
        # do something
case "matrix":
    _data = data
    if data.dtype.kind not in "iuf":
    # do something
```
or 
```
case "arg" | "file" | "geojson" | "grid" | "stringio":
    _data = data
case "image" if data.dtype != "uint8":
    raise ...
case "matrix" if data.dtype.kind not in "iuf":
    ...
case _:
    _data = data # this is necessary to catch cases like kind="matrix" and dtype in "iuf"  
```

I still prefer the current codes, which set `_data` to `data` by default and the match-case statements only need to deal with exceptions.

> So that if there's a new `kind` in the future, we know to add it explicitly.

If there is a new `kind` in the future, `_data=data` is still the default. If needs special handling and we forget to add it, then the related tests will fail anyway.

Reply:
Ah ok, that makes sense, let's keep `_data = data` at the top then.